lib/bugsnag/helpers.rb

```diff
@@ -60,21 +70,74 @@ def self.deep_merge!(l_hash, r_hash)
 
     TRUNCATION_INFO = '[TRUNCATED]'
 
+    ##
+    # Trim stacktrace code out if they're too large, oldest functions first
+    def self.trim_stacktrace_code(payload, threshold)
+      extract_exception(payload) do |exception|
+        total = exception[:stacktrace].length
+        count = 0
+        saved = 0
+        while (count < total) && (saved < threshold)
+          trace = exception[:stacktrace][total - count - 1]
+          saved += get_payload_length(trace[:code])
+          trace.delete(:code)
+          count += 1
+        end
+      end
+      payload
+    end
+
+    ##
+    # Trim stacktrace entries out oldest functions first
+    def self.trim_stacktrace_functions(payload, threshold)
+      extract_exception(payload) do |exception|
+        saved = 0
+        while (exception[:stacktrace].size > 1) && (saved < threshold)
+          saved += get_payload_length(exception[:stacktrace].last)
+          exception[:stacktrace].pop
+        end
+      end
+      payload
+    end
+
+    ##
+    # Wrapper for trimming stacktraces
+    def self.extract_exception(payload)
+      valid_payload = payload.is_a?(Hash) && payload[:events].respond_to?(:map)
+      return unless valid_payload && block_given?
+      payload[:events].each do |event|
+        event[:exceptions].each { |exception| yield exception }
+      end
+    end
+
+    ##
+    # Take the metadata from the events and trim it down
+    def self.trim_metadata(payload)
+      return payload unless payload.is_a?(Hash) and payload[:events].respond_to?(:map)
+      payload[:events].map do |event|
+        event[:metaData] = truncate_arrays_in_value(event[:metaData])
+        event[:metaData] = trim_strings_in_value(event[:metaData])
+      end
+      payload
+    end
+
     ##
     # Check if a value is a raw type which should not be trimmed, truncated
     # or converted to a string
     def self.is_json_raw_type?(value)
       RAW_DATA_TYPES.detect {|klass| value.is_a?(klass)} != nil
     end
 
+    ##
     # Shorten array until it fits within the payload size limit when serialized
-    def self.truncate_array(array)
+    def self.truncate_array(array, limit = MAX_ARRAY_LENGTH)
```

Comment:
If `limit` is just a constant value everywhere and never modified, just use it directly instead of passing it from one method to the next.